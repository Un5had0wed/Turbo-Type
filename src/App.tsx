import { useReducer } from "react";
import { AppStates, AppStateActions } from "./components/types/App.type";
import Form from "./components/Form";
import GameBoard from "./components/GameBoard";
import SetWordlist from "./assets/service/SetWordlist";
import "./assets/styles/App.css";

const initialStates: AppStates = {
  pageMode: "Home",
  userName: "",
  level: "Easy",
  dFactor: 1,
  wordList: SetWordlist("Easy"),
  randNum: Math.floor(Math.random() * SetWordlist("Easy").length),
};

function reducer(state: AppStates, action: AppStateActions) {
  switch (action.type) {
    case "setPageMode":
      return {
        ...state,
        pageMode: action.payload as "Home" | "Game",
      };
    case "setName":
      return {
        ...state,
        userName: action.payload,
      };
    case "setLevel":
      return {
        ...state,
        level: action.payload as "Easy" | "Medium" | "Hard",
        dFactor:
          action.payload === "Hard" ? 2 : action.payload === "Medium" ? 1.5 : 1,
        wordList: SetWordlist(action.payload),
        randNum: Math.floor(Math.random() * SetWordlist(action.payload).length),
      };
    default:
      return state;
  }
}

function App() {
  const [appStates, dispatch] = useReducer(reducer, initialStates);
  if (appStates.pageMode === "Home") {
    return (
      <div className="app">
        <Form appStates={appStates} setAppState={dispatch} />
      </div>
    );
  } else if (appStates.pageMode === "Game") {
    return (
      <div className="app">
        <GameBoard appStates={appStates} />
      </div>
    );
  }
}

export default App;
