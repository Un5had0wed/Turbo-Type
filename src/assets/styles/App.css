* {
    padding: 0;
    margin: 0;
}
  
#root {
    width: 100%;
    height: 100vh;
    background-image: url("../images/background.jpg");
    background-size: cover;
    background-color: black;
    background-position: center;
}
  