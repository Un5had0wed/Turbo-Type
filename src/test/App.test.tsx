import { render, screen, fireEvent } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect'; // for toBeInTheDocument
import App from '../App';

// Mock SetWordlist
jest.mock('../assets/service/SetWordlist', () => ({
  __esModule: true,
  default: jest.fn(() => ['mocked', 'wordlist']),
}));

describe('App Component', () => {
  test('renders App component with Home page mode', () => {
    render(<App />);
    expect(screen.getByText('Typing Triumph Challenge')).toBeInTheDocument();
  });

  test('renders App component with Game page mode', () => {
    render(<App />);
    fireEvent.change(screen.getByPlaceholderText('Enter your name'), {
      target: { value: 'John Doe' },
    });
    fireEvent.click(screen.getByText('Start Game'));
    expect(screen.getByText('SCORE BOARD')).toBeInTheDocument();
  });

  test('reducer sets page mode correctly', () => {
    const { container } = render(<App />);
    fireEvent.click(screen.getByText('Start Game'));
    expect(container.firstChild).toMatchSnapshot(); // Check if component snapshot matches
  });

  test('reducer sets user name correctly', () => {
    const { container } = render(<App />);
    fireEvent.change(screen.getByPlaceholderText('Enter your name'), {
      target: { value: 'John Doe' },
    });
    expect(container.firstChild).toMatchSnapshot();
  });

  test('reducer sets level and dFactor correctly', () => {
    const { container } = render(<App />);
    fireEvent.click(screen.getByText('Easy'), {
      target: { value: 'Medium' },
    });
    expect(container.firstChild).toMatchSnapshot();
  });
});
