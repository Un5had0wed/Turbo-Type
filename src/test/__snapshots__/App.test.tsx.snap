// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App Component reducer sets level and dFactor correctly 1`] = `
<div
  class="app"
>
  <div
    class="form"
  >
    <form>
      <img
        alt="logo"
        class="form__logo"
        src="test-file-stub"
      />
      <div
        class="heading__grp"
      >
        <h1
          class="form__heading form__heading--main"
        >
          Turbo Type
        </h1>
        <span
          class="form__heading form__heading--tagline"
        >
          Typing Triumph Challenge
        </span>
      </div>
      <input
        class="form__inp"
        name="name"
        placeholder="Enter your name"
        required=""
        type="text"
        value=""
      />
      <select
        class="form__select"
        name="level"
      >
        <option
          value="Medium"
        >
          Easy
        </option>
        <option
          value="Medium"
        >
          Medium
        </option>
        <option
          value="Hard"
        >
          Hard
        </option>
      </select>
      <button
        class="form__btn"
        type="submit"
      >
        Start Game
      </button>
    </form>
  </div>
</div>
`;

exports[`App Component reducer sets page mode correctly 1`] = `
<div
  class="app"
>
  <div
    class="form"
  >
    <form>
      <img
        alt="logo"
        class="form__logo"
        src="test-file-stub"
      />
      <div
        class="heading__grp"
      >
        <h1
          class="form__heading form__heading--main"
        >
          Turbo Type
        </h1>
        <span
          class="form__heading form__heading--tagline"
        >
          Typing Triumph Challenge
        </span>
      </div>
      <input
        class="form__inp"
        name="name"
        placeholder="Enter your name"
        required=""
        type="text"
        value=""
      />
      <select
        class="form__select"
        name="level"
      >
        <option
          value="Easy"
        >
          Easy
        </option>
        <option
          value="Medium"
        >
          Medium
        </option>
        <option
          value="Hard"
        >
          Hard
        </option>
      </select>
      <button
        class="form__btn"
        type="submit"
      >
        Start Game
      </button>
    </form>
  </div>
</div>
`;

exports[`App Component reducer sets user name correctly 1`] = `
<div
  class="app"
>
  <div
    class="form"
  >
    <form>
      <img
        alt="logo"
        class="form__logo"
        src="test-file-stub"
      />
      <div
        class="heading__grp"
      >
        <h1
          class="form__heading form__heading--main"
        >
          Turbo Type
        </h1>
        <span
          class="form__heading form__heading--tagline"
        >
          Typing Triumph Challenge
        </span>
      </div>
      <input
        class="form__inp"
        name="name"
        placeholder="Enter your name"
        required=""
        type="text"
        value="John Doe"
      />
      <select
        class="form__select"
        name="level"
      >
        <option
          value="Easy"
        >
          Easy
        </option>
        <option
          value="Medium"
        >
          Medium
        </option>
        <option
          value="Hard"
        >
          Hard
        </option>
      </select>
      <button
        class="form__btn"
        type="submit"
      >
        Start Game
      </button>
    </form>
  </div>
</div>
`;
